modules/substrate/src/lib.rs

```diff
@@ -371,27 +375,93 @@ impl<T: Config> bp_header_chain::HeaderChain<BridgedHeader<T>> for Module<T> {
 		PalletStorage::<T>::new().current_authority_set()
 	}
 
-	fn import_header(header: BridgedHeader<T>) -> Result<(), ()> {
-		let mut verifier = verifier::Verifier {
-			storage: PalletStorage::<T>::new(),
-		};
+	fn append_finalized_chain(
+		headers: impl IntoIterator<Item = BridgedHeader<T>>,
+	) -> Result<(), sp_runtime::DispatchError> {
+		let mut storage = PalletStorage::<T>::new();
+
+		let mut header_iter = headers.into_iter().peekable();
+		let first_header = header_iter.peek().ok_or(Error::<T>::NotDescendant)?;
+		let best_finalized = storage.best_finalized_header().header;
+
+		// Quick ancestry check to make sure we're not writing complete nonsense to storage
+		ensure!(
+			best_finalized.hash() == *first_header.parent_hash(),
```

Comment:
Not for this PR, but probably you should replace `storage.best_finalized_header().header.hash()` with something like `storage.best_finalized_header_hash()` - you'll only need single read in this case, instead of two reads + decode + hash.

Reply:
Oh good point, I forgot that we had that available as a storage item